File: rust/kernel-receiver/src/display/log.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};

use crate::config::Config;
use crate::stats::Stats;

fn format_signature_prefix(sig: &[u8; 8]) -> String {
    let hex: String = sig.iter().map(|b| format!("{b:02x}")).collect();
    format!("{}..{}", &hex[..4], &hex[12..])
}

pub fn run(
    config: &Config,
    stats: Arc<RwLock<Stats>>,
    shutdown: Arc<AtomicBool>,
) -> anyhow::Result<()> {
    let interval = Duration::from_secs(config.display.log_interval_secs);
    let mut last_print = Instant::now();
    let mut last_reported_slots: Vec<u64> = Vec::new();

    eprintln!(
        "Log mode: printing stats every {}s. Press Ctrl+C to stop.",
        config.display.log_interval_secs
    );

    while !shutdown.load(Ordering::Relaxed) {
        std::thread::sleep(Duration::from_millis(100));

        if last_print.elapsed() < interval {
            continue;
        }
        last_print = Instant::now();

        let mut s = stats.write().unwrap();

        // Print per-slot lines for newly seen slots
        let current_slots: Vec<u64> = s.slots.keys().copied().collect();
        for &slot_num in &current_slots {
            if last_reported_slots.contains(&slot_num) {
                continue;
            }
            if let Some(slot) = s.get_slot(slot_num) {
                let age_ms = slot.first_seen.elapsed().as_millis();
                println!(
                    "slot={} sig={} data={} coding={} fec_sets={} age_ms={}",
                    slot.slot,
                    format_signature_prefix(&slot.signature_prefix),
                    slot.data_shred_count,
                    slot.coding_shred_count,
                    slot.fec_set_count,
                    age_ms,
                );
            }
        }
        last_reported_slots = current_slots;

        // Print summary line
        let rate = s.shreds_per_second();
        let hb_ago = s
            .last_heartbeat
            .map(|t| format!("{}ms ago", t.elapsed().as_millis()))
            .unwrap_or_else(|| "never".into());

        println!(
            "[stats] shreds/sec={:.0} data={} coding={} errors={} heartbeats={} (last: {})",
            rate,
            s.total_data_shreds,
            s.total_coding_shreds,
            s.parse_errors,
            s.total_heartbeats,
            hb_ago,
        );
    }

    Ok(())
}

File: go/internal/display/log.go
package display

import (
	"context"
	"fmt"
	"os"
	"sync"
	"time"

	"github.com/malbeclabs/edge-multicast-ref/go/internal/stats"
)

// RunLogDisplay periodically prints shred stats to stderr until the context
// is cancelled. It sleeps in 100ms increments to allow prompt shutdown, and
// prints a report every intervalSecs seconds. Only new (previously unreported)
// slots are printed each cycle.
func RunLogDisplay(ctx context.Context, s *stats.Stats, mu *sync.RWMutex, intervalSecs int) {
	interval := time.Duration(intervalSecs) * time.Second
	lastReport := time.Now()
	reportedSlots := make(map[uint64]struct{})
	var lastTotalShreds uint64

	for {
		select {
		case <-ctx.Done():
			return
		default:
		}

		time.Sleep(100 * time.Millisecond)

		elapsed := time.Since(lastReport)
		if elapsed < interval {
			continue
		}
		lastReport = time.Now()

		mu.Lock()

		// Print new slot lines.
		recent := s.RecentSlots()
		for _, ss := range recent {
			if _, seen := reportedSlots[ss.Slot]; seen {
				continue
			}
			reportedSlots[ss.Slot] = struct{}{}

			ageMs := time.Since(ss.FirstSeen).Milliseconds()
			sigStr := FormatSignaturePrefix(ss.SignaturePrefix)
			fmt.Fprintf(os.Stderr,
				"slot=%d sig=%s data=%d coding=%d fec_sets=%d age_ms=%d\n",
				ss.Slot, sigStr,
				ss.DataShredCount, ss.CodingShredCount,
				ss.FECSetCount, ageMs,
			)
		}

		// Compute rate from delta since last report.
		totalNow := s.TotalDataShreds + s.TotalCodingShreds
		elapsedSecs := elapsed.Seconds()
		var rate float64
		if elapsedSecs > 0 {
			rate = float64(totalNow-lastTotalShreds) / elapsedSecs
		}
		lastTotalShreds = totalNow
		lastHBStr := "never"
		if s.LastHeartbeat != nil {
			lastHBMs := time.Since(*s.LastHeartbeat).Milliseconds()
			lastHBStr = fmt.Sprintf("%dms ago", lastHBMs)
		}
		fmt.Fprintf(os.Stderr,
			"[stats] shreds/sec=%.0f data=%d coding=%d errors=%d heartbeats=%d (last: %s)\n",
			rate,
			s.TotalDataShreds, s.TotalCodingShreds,
			s.ParseErrors, s.TotalHeartbeats,
			lastHBStr,
		)

		mu.Unlock()
	}
}
